### PR TITLE
fix(indexer): route archive-depth eth_call failures to fallback RPC

### PR DESCRIPTION
**Pull request by chapati23**
## Summary

- Adds an archive-depth recovery branch to `readContractWithBlockFallback` so historical `eth_call`s that the primary RPC can't serve (because its archive depth doesn't reach back that far) get routed to the **secondary RPC at the same block** instead of degrading to `latest` or just returning null.
- Caught in PR #329's deploy logs: 3,710 `fetchBreakerList` + 2,780 `reportExpiry` + ~1,000 each `getRebalancingState` / `getTradingLimits` failures during catch-up — all on Monad, all the same root cause (quiknode lacks archive ~5M blocks back from head).
- The codebase already has a fallback RPC wired for rate-limit handling (`getFallbackRpcClient`) — \`rpc2.monad.xyz\` for Monad — and verified via \`cast\` probes that **it does have the archive depth at the failing blocks**. The fallback was just never consulted on archive-depth errors. Now it is.

## Why now

- PR #331's chain-time TTL turned out to be a much weaker bandage than I estimated (deploy logs show the storm only modestly reduced — see the analysis on PR #331's thread).
- Routing through a deeper-archive fallback fixes the root cause: every historical `eth_call` that quiknode can't serve gets retried on \`rpc2.monad.xyz\` at the SAME block, preserving block-scoped accuracy.
- All four storm sources (\`fetchBreakerList\`, \`reportExpiry\`, \`getRebalancingState\`, \`getTradingLimits\`) collapse together since they all flow through \`readContractWithBlockFallback\`.

## Recovery flow after this PR

1. Primary fails with archive-depth error → try **secondary at same block**
2. Secondary returns data → \`usedFallback=true\`, \`usedLatestFallback=false\` (block-scoped, NO accuracy loss)
3. Secondary also fails → fall through to reading \`latest\` on primary (existing behaviour, marks \`usedLatestFallback=true\` so block-scoped callers can discard)

Skips the retry loop for archive-depth — those would deterministically fail with the same error against the same primary, wasting backoff time.

## What this DOESN'T do

- HyperRPC: still NOT a candidate for \`eth_call\` (no support, see \`src/rpc/client.ts:140-142\` and the envio skill memory).
- Rate-limit handling: unchanged. Existing path (try primary with retries → fall back to secondary on persistent rate-limit) is untouched.
- PR #331's chain-time TTL: kept in place as defense-in-depth (bounded via FIFO eviction, no harm) but its primary motivation is addressed at the root here.

## Test plan

- [x] \`pnpm typecheck\` clean
- [x] \`pnpm lint\` clean
- [x] \`pnpm mocha\` — 538 passing (4 new):
  - \`archive-depth: secondary at SAME block returns block-scoped result\`
  - \`archive-depth: matches 'querying historical state' phrasing too\`
  - \`archive-depth: secondary also fails → falls through to \`latest\` on primary\`
  - \`archive-depth: no fallback configured → straight to \`latest\`, no retries\`
- [ ] After merge + deploy, verify burst counters for \`fetchBreakerList\` / \`reportExpiry\` / \`getRebalancingState\` / \`getTradingLimits\` on Monad collapse from thousands → near-zero. Watch for any new \`[RPC_RATE_LIMIT_FALLBACK]\` warnings — if \`rpc2.monad.xyz\` rate-limits us under the new traffic, that's the next thing to handle.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes core `readContractWithBlockFallback` error routing to classify archive-depth vs transient block-miss and to consult the secondary RPC at the requested block; misclassification could affect historical data correctness or cause more null/throws during catch-up.
> 
> **Overview**
> Improves historical `eth_call` reliability by adding an **archive-depth** branch to `readContractWithBlockFallback`: when the primary RPC reports "querying historical state" it now retries the **fallback RPC at the same block** (and *never* degrades to `latest` for this case), with sanitized secondary-error logging.
> 
> Extends block-miss detection to include bare "block requested not found", tightens interaction with rate-limit retries (only route retry-surfaced *archive-depth* errors to the new path; other post-retry block-misses fail closed), and updates multiple breakers/pool-state fetchers to **reject `usedLatestFallback` results** to avoid persisting wrong-block governance/state data.
> 
> Adds comprehensive unit tests covering archive-depth fallback behavior, error sanitization, and rate-limit/regex edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 206f839c0d3d237c20fc9290e1709e9335ddb04c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by chapati23** [comment]
@claude review

**Comment by chapati23** [comment]
cursor review

File: indexer-envio/src/rpc/block-fallback.ts [hunk 1]
```diff
@@ -21,6 +21,19 @@ import {
...
+ * doesn't have state pruned that far back*. Different providers:
+ * - "Block requested not found" (quiknode)
+ * - "querying historical state that is not available" (quiknode/others)
+ *
+ * Distinct from BLOCK_NOT_AVAILABLE_RE (which means "the chain hasn't
+ * reached this block yet"). Archive-depth failures are recoverable via a
+ * secondary RPC with deeper archive at the SAME block — block-scoped
+ * accuracy is preserved. BLOCK_NOT_AVAILABLE_RE failures are recoverable
+ * only by retrying or reading `latest` (different block). */
+const ARCHIVE_DEPTH_RE = /block requested not found|querying historical state/i;
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts]: Fixed in 0b680df. Narrowed `ARCHIVE_DEPTH_RE` from `/block requested not found|querying historical state/i` to just `/querying historical state/i` — the unambiguous archive-depth marker. Quiknode's actual error includes both phrases together ("Block requested not found. Request might be querying historical state...") so we still match production. Bare "Block requested not found" without the historical-state qualifier (which other providers may emit for transient lag) now falls through to the existing BLOCK_NOT_AVAILABLE_RE retry path. New test `archive-depth regex: bare 'block requested not found' (no historical-state qualifier) does NOT trigger archive-depth path` pins this.

File: indexer-envio/src/rpc/block-fallback.ts [hunk 2]
```diff
@@ -131,7 +144,60 @@ export async function readContractWithBlockFallback(
...
+        // Secondary also failed (rate-limit, deeper-archive miss, or some
+        // other transient). Fall through to the block-not-available
+        // "read latest" path below — both BLOCK_NOT_AVAILABLE_RE and
+        // ARCHIVE_DEPTH_RE patterns trigger that branch, so the original
+        // primary error (`err`) keeps the flow moving rather than the
+        // secondary error short-circuiting it.
+        console.warn(
+          `[RPC_ARCHIVE_FALLBACK_FAILED] fn=${fn} target=${target} requestedBlock=${blockNumber} — both primary and secondary failed; falling through to latest-block fallback`,
+        );
+      }
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts]: Fixed in 0b680df. The fail-closed fix subsumes this: secondary errors are now thrown verbatim (`throw fallbackErr`), not converted to a `latest` read. `fetchRebalanceIncentiveAtBlock` will see the original "returned no data" message and stamp the -2 sentinel correctly. New test `archive-depth: secondary returns 'returned no data' → throws so caller classifies it` pins this.

File: indexer-envio/src/rpc/block-fallback.ts [hunk 3]
```diff
@@ -131,7 +144,60 @@ export async function readContractWithBlockFallback(
...
+      blockNumber !== undefined &&
+      err instanceof Error &&
+      ARCHIVE_DEPTH_RE.test(err.message)
+    ) {
+      console.warn(
+        `[RPC_BLOCK_FALLBACK] fn=${fn} target=${target} requestedBlock=${blockNumber} — primary lacks archive depth and no usable fallback, reading latest instead`,
+      );
+      const result = await makeCall(client, undefined);
+      return { result, usedFallback: true, usedLatestFallback: true };
+    }
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts:257]: Fixed in 0b680df — fail-closed instead of silently feeding `latest` data to historical callers. Effects' batch memoization no longer amplifies wrong-block data because `usedLatestFallback=true` is never set on the archive-depth path. Threw out the `latest`-fallback branch entirely; secondary failure or absent fallback throws.

File: indexer-envio/src/rpc/block-fallback.ts [hunk 4]
```diff
@@ -131,7 +144,60 @@ export async function readContractWithBlockFallback(
...
+    if (
+      blockNumber !== undefined &&
+      err instanceof Error &&
+      ARCHIVE_DEPTH_RE.test(err.message)
+    ) {
+      console.warn(
+        `[RPC_BLOCK_FALLBACK] fn=${fn} target=${target} requestedBlock=${blockNumber} — primary lacks archive depth and no usable fallback, reading latest instead`,
+      );
+      const result = await makeCall(client, undefined);
+      return { result, usedFallback: true, usedLatestFallback: true };
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts]: Fixed in 0b680df — same fix as the other thread. Archive-depth path no longer reads `latest` on secondary failure or missing fallback; it throws and lets each caller's existing try/catch return null. None of the historical-block consumers (`fetchBreakerList`, `fetchReportExpiry`, `fetchTradingLimits`, etc.) need to start checking `usedLatestFallback` because we never set it to true on this path anymore.

[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts]: Fixed in 0b680df. Made archive-depth fail-closed: secondary failure or no fallback now throws to the caller's existing try/catch instead of falling through to `latest`. Restores pre-PR behaviour (callers degrade to null) and avoids the silent corruption of historical entity state. The `latest`-fallback path for archive-depth is gone entirely.

**Comment by chapati23** [comment]
@claude — addressed all 3 notes in 0b680df. Summary:

- **#1 Secondary failure not logged with message** → Fixed in 0b680df. `[RPC_ARCHIVE_FALLBACK_FAILED]` now includes `secondaryErr="<message>"`. When `rpc2.monad.xyz` eventually rate-limits or returns an unfamiliar error under the new traffic pattern, the log line will show which.

- **#2 `[RPC_BLOCK_FALLBACK]` tag reused** → Fixed in 0b680df, but indirectly. The Codex/Cursor reviews flagged that the archive-depth → `latest` path was a regression (multiple historical callers don't check `usedLatestFallback`). I fixed that by making archive-depth fail-closed entirely — no `latest` read on this path anymore — which deletes the duplicate-tag concern as a side-effect. The `[RPC_BLOCK_FALLBACK]` tag is now used only by the BLOCK_NOT_AVAILABLE retry path.

- **#3 Two ARCHIVE_DEPTH_RE checks could confuse readers** → Fixed in 0b680df. Same root cause as #2 — there's now only one ARCHIVE_DEPTH_RE branch, fail-closed.

Codex + Cursor independently flagged the same correctness regression you noted as an "edge case worth knowing" (returning `latest` data to historical callers). All three of those threads converge on the same fix: archive-depth is now fail-closed, throws on secondary failure or missing fallback, callers' existing try/catch returns null. Pre-PR behaviour preserved on the failure path.

Tests: 540 passing (5 archive-depth cases now: secondary succeeds, alternative phrasing, secondary fails throws, no fallback throws, bare "block requested not found" stays unrecognised), typecheck + lint clean.

File: indexer-envio/src/rpc/block-fallback.ts [hunk 5]
```diff
@@ -131,7 +156,65 @@ export async function readContractWithBlockFallback(
...
+          // Throw the secondary's error so callers can classify it
+          // correctly — e.g. fetchRebalanceIncentiveAtBlock needs to see
+          // "returned no data" to stamp the -2 sentinel for older pools
+          // without the getter.
+          const secondaryMsg =
+            fallbackErr instanceof Error
+              ? fallbackErr.message
+              : String(fallbackErr);
+          console.warn(
+            `[RPC_ARCHIVE_FALLBACK_FAILED] fn=${fn} target=${target} requestedBlock=${blockNumber} secondaryErr="${secondaryMsg}" — propagating to caller`,
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts]: Fixed in 044f0db. Exported `sanitizeErrorMessage` from `src/rpc/client.ts` (was internal — used by `logRpcFailure` for URL redaction) and now wrap `secondaryMsg` with it before logging. Tokenized URLs (HyperRPC's `*/<token>`, quiknode-with-token endpoints) get redacted to `<host>/<redacted>` before they hit the warn line. New test `archive-depth: secondary error sanitization redacts URL-bearing messages before logging` captures stderr and asserts the token doesn't appear in the log.

File: indexer-envio/src/rpc/block-fallback.ts [hunk 6]
```diff
@@ -131,7 +156,65 @@ export async function readContractWithBlockFallback(
...
+    // src/rpc/{breakers,pool-state}.ts — destructure `result` directly
+    // without checking `usedLatestFallback`, so silently swapping in
+    // current-block data would corrupt historical entity state. Throwing
+    // is the same behaviour those callers had before this PR (their
+    // try/catch returns null), so fail-closed is the safe default.
+    if (
+      blockNumber !== undefined &&
+      err instanceof Error &&
+      ARCHIVE_DEPTH_RE.test(err.message)
+    ) {
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts:224]: Fixed in 044f0db. Hoisted the catch-bound error to a reassignable `currentError`. When the rate-limit retry loop surfaces a non-rate-limit error (e.g. archive-depth after the rate-limit cleared), we now `break` out of the loop with `currentError = retryErr` instead of throwing immediately. The archive-depth / block-not-available branches below then classify it the same way they would an initial primary error — so the same-block secondary fallback IS consulted on the post-retry archive-depth path. New test `archive-depth: error surfaced AFTER rate-limit retry routes to secondary at same block` pins this.

File: indexer-envio/src/rpc/block-fallback.ts [hunk 7]
```diff
@@ -106,36 +141,110 @@ export async function readContractWithBlockFallback(
...
+          if (!isRateLimitError(retryErr)) {
+            // Non-rate-limit error surfaced during retry. Reassign and let
+            // the archive-depth / block-not-available branches below
+            // classify it — otherwise an archive-depth error that only
+            // emerged after the rate-limit cleared would skip the
+            // same-block secondary fallback and bubble straight to the
+            // caller.
+            currentError = retryErr;
+            exitedWithRateLimit = false;
+            break;
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts]: Fixed in d18678a. Narrowed the post-rate-limit-retry routing to ARCHIVE_DEPTH_RE only. Block-miss errors surfacing late now throw fail-closed (matches pre-PR behaviour for the same retry-surfaced shape) — historical callers `fetchNumReporters`, `fetchReportExpiry`, `fetchTradingLimits` won't see `usedLatestFallback=true` from this path. Archive-depth surfacing late still routes to the same-block secondary because that's the only path where block-scoped accuracy CAN be preserved. New test: `post-rate-limit-retry block-miss: throws fail-closed (does NOT route to retry-then-latest)`, plus a companion regression guard `post-rate-limit-retry archive-depth: still routes to secondary`.

File: indexer-envio/test/blockFallback.test.ts [hunk 8]
```diff
@@ -282,4 +282,228 @@ describe("readContractWithBlockFallback", () => {
...
+    await assert.rejects(
+      readContractWithBlockFallback(primary, baseArgs, 100n, null),
+      /Block requested not found/,
+      "bare 'block requested not found' is treated as an unrecognised error, not archive-depth",
+    );
+    assert.equal(
+      primaryCallNo,
+      1,
+      "no retries / latest-fallback for unrecognised errors",
+    );
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/test/blockFallback.test.ts:548]: Fixed in d18678a. You're right — my regex comment was aspirational but the regex itself didn't include the bare phrase. Added `block requested not found` to `BLOCK_NOT_AVAILABLE_RE`. Combined phrasing (the "...querying historical state..." suffix quiknode emits) still routes to ARCHIVE_DEPTH_RE first because the dispatcher checks that branch first. Bare phrasing from any other provider now retries-then-latest like the other transient-lag patterns. Updated test `archive-depth regex: bare 'block requested not found' is treated as transient-lag, NOT archive-depth` to match the new behaviour.

File: indexer-envio/src/rpc/block-fallback.ts [hunk 9]
```diff
@@ -10,16 +10,48 @@ import {
...
+ * - "unknown block" (Nethermind)
+ * - "block requested not found" (some providers, when used WITHOUT the
+ *   "querying historical state" qualifier — bare phrasing tends to mean
+ *   transient lag rather than archive depth). When the same string co-
+ *   occurs with "querying historical state" we treat it as archive-depth
+ *   instead via ARCHIVE_DEPTH_RE; the dispatcher checks ARCHIVE_DEPTH_RE
+ *   first and never falls through to here for that combined phrasing. */
 const BLOCK_NOT_AVAILABLE_RE =
-  /block is out of range|block number out of range|header not found|unknown block/i;
+  /block is out of range|block number out of range|header not found|unknown block|block requested not found/i;
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts:29]: Fixed structurally in ba27df4 — addressed at the call sites rather than the dispatcher. Each round of dispatcher fixes was finding a new corner where `latest`-block data could leak through; the root cause is that 5 fetchers (`fetchNumReporters`, `fetchReportExpiry`, `fetchTradingLimits`, `fetchBreakerList`, `fetchBreakerDefaults`, `fetchBreakerFeedState`) consumed `result` without checking `usedLatestFallback`. Now they all reject latest-fallback returns at the fetcher boundary, so this concern is moot regardless of which dispatcher branch fired. `fetchRebalancingState` / `fetchReserves` / `fetchRebalanceIncentiveAtBlock` already did this — after this commit the check is universal across all block-scoped fetchers. The dispatcher's round-1/2/3 fixes remain correct on their own merits but are no longer load-bearing for safety.

[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts:29]: Fixed structurally in ba27df4 — addressed at the call sites rather than the dispatcher. Each round of dispatcher fixes was finding a new corner where `latest`-block data could leak through; the root cause is that 5 fetchers (`fetchNumReporters`, `fetchReportExpiry`, `fetchTradingLimits`, `fetchBreakerList`, `fetchBreakerDefaults`, `fetchBreakerFeedState`) consumed `result` without checking `usedLatestFallback`. Now they all reject latest-fallback returns at the fetcher boundary, so this concern is moot regardless of which dispatcher branch fired. `fetchRebalancingState` / `fetchReserves` / `fetchRebalanceIncentiveAtBlock` already did this — after this commit the check is universal across all block-scoped fetchers. The dispatcher's round-1/2/3 fixes remain correct on their own merits but are no longer load-bearing for safety.

[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts:29]: Fixed structurally in ba27df4 — addressed at the call sites rather than the dispatcher. Each round of dispatcher fixes was finding a new corner where `latest`-block data could leak through; the root cause is that 5 fetchers (`fetchNumReporters`, `fetchReportExpiry`, `fetchTradingLimits`, `fetchBreakerList`, `fetchBreakerDefaults`, `fetchBreakerFeedState`) consumed `result` without checking `usedLatestFallback`. Now they all reject latest-fallback returns at the fetcher boundary, so this concern is moot regardless of which dispatcher branch fired. `fetchRebalancingState` / `fetchReserves` / `fetchRebalanceIncentiveAtBlock` already did this — after this commit the check is universal across all block-scoped fetchers. The dispatcher's round-1/2/3 fixes remain correct on their own merits but are no longer load-bearing for safety.

File: indexer-envio/src/rpc/block-fallback.ts [hunk 10]
```diff
@@ -106,36 +147,123 @@ export async function readContractWithBlockFallback(
...
+    // src/rpc/{breakers,pool-state}.ts — destructure `result` directly
+    // without checking `usedLatestFallback`, so silently swapping in
+    // current-block data would corrupt historical entity state. Throwing
+    // is the same behaviour those callers had before this PR (their
+    // try/catch returns null), so fail-closed is the safe default.
+    if (
+      blockNumber !== undefined &&
+      currentError instanceof Error &&
+      ARCHIVE_DEPTH_RE.test(currentError.message)
+    ) {
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/block-fallback.ts:224]: Fixed structurally in ba27df4 — addressed at the call sites rather than the dispatcher. Each round of dispatcher fixes was finding a new corner where `latest`-block data could leak through; the root cause is that 5 fetchers (`fetchNumReporters`, `fetchReportExpiry`, `fetchTradingLimits`, `fetchBreakerList`, `fetchBreakerDefaults`, `fetchBreakerFeedState`) consumed `result` without checking `usedLatestFallback`. Now they all reject latest-fallback returns at the fetcher boundary, so this concern is moot regardless of which dispatcher branch fired. `fetchRebalancingState` / `fetchReserves` / `fetchRebalanceIncentiveAtBlock` already did this — after this commit the check is universal across all block-scoped fetchers. The dispatcher's round-1/2/3 fixes remain correct on their own merits but are no longer load-bearing for safety.
